#31793 Add support for subtracting datetime from Timestamp

AnjoMan opened the pull request:
- [x] closes #31793
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Based on the discussion in the attached issue, i've made the following changes:

1. add support for subtracting when the two items have different timezones
2. raise an error if one (but not both) timezone is naive

The root cause of 31793 was that the tzinfo objects were from different libraries -- `Timestamp` makes a `pytz.UTC` whereas a datetime object might just have `datetime.timezone.utc`. These objects are not necessarily equal, even if they represent the same timezone.

In addition, this change *adds* support for time-deltas where the timezone of each date is not the same. This follows the behaviour of `datetime` more closely.


I reference the [python `datetime` module documentation](https://docs.python.org/3/library/datetime.html) which defines when a datetime is naive: when its `tzinfo` is `None` or a call to `item.tzinfo.utcoffset(item)` returns `None`.

AnjoMan commented:
If this is a welcome change, please add the `hacktoberfest-accepted` to this issue so it can go towards my event contribution

AnjoMan commented:
Ok, looking at the failed tests, it l;ooks like i should be relocating the tests i added so they would appear under `tests/arithmetic/test_timedetlta64.py` and `tests/scalar/timestamp/test_arithmetic.py`.

arw2019 commented on pandas/_libs/tslibs/timestamps.pyx:
```diff
@@ -368,9 +368,13 @@ cdef class _Timestamp(ABCTimestamp):
-            if not tz_compare(self.tzinfo, other.tzinfo):
-                raise TypeError("Timestamp subtraction must have the "
-                                "same timezones or no timezones")
+            naive_timezones = [
```
personal taste but I find something like
```
(self.tzinfo is None and other.tzinfo is not None) or (self.tzinfo is not None and other.tzinfo is None)
```
easier to read here than the list comp and any/all

jbrockmendel replied:
more concise: `(self.tzinfo is None) ^ (other.tzinfo is None)`

AnjoMan replied:
wondering if we need to handle the other case mentioned [here](https://docs.python.org/3/library/datetime.html#determining-if-an-object-is-aware-or-naive), where tzinfo is not `None` but doing `time.tzinfo.utcoffset(time)` returns None? i'm not really aware of how common such a case is but it also seems like there isn't much cost to including it other than the way the code will be written. 

if we were to support that case (what i have currently does), i would probably use a function like
```
def is_naive(time):
     return time.tzinfo is None or time.tzinfo.utcoffset(time) is None
```

This short-circuits on the more likely case that tzinfo is None, but also covers the return value of `utcoffset`

AnjoMan replied:
ok, i switched to the symmetric inequality operator formulation `^`. thanks @jbrockmendel for the suggestion

arw2019 commented on pandas/_libs/tslibs/timestamps.pyx:
```diff
@@ -368,9 +368,13 @@ cdef class _Timestamp(ABCTimestamp):
+                for el in (self, other)
+            ]
+            if any(naive_timezones) and not all(naive_timezones):
+                raise TypeError("Cannot compute time difference between"
```
I'd do: `Timedelta between Timestamps with different timezone object types is not supported.`

jbrockmendel replied:
could use assert_tzawareness_compat

AnjoMan replied:
i don't think the suggested error message is precise enough, the key issue here is whether you can ascertain the timezone of both dates with confidence:

* if both are naive, we can assume they are both in local time; the time difference doesn't depend on the timezone
* if both are timezone-aware, we can compute the absolute time difference (e.g. 12pm EST - 12PM PT is 3 hrs; 12pm EST - 9AM PT is 0 hrs)
* if only one is naive, we have to fail because we can't know what the timezone offset is between the naive and non-naive times

This also makes me think that we should be raising `ValueError` here instead of `TypeError`.

arw2019 commented on pandas/tests/tslibs/test_timedeltas.py:
```diff
@@ -34,3 +34,41 @@ def test_huge_nanoseconds_overflow():
+        with pytest.raises(TypeError, match=msg):
+            time2 - time1
+
+    def test_subtracts_dates_of_different_timezones(self):
```
These aren't different *timezones*, are they?

AnjoMan replied:
the timezone offsets are different, one is `04:00` and the other is `00:00`. i could change the name to offsets to make it more concise

arw2019 commented on pandas/tests/tslibs/test_timedeltas.py:
```diff
@@ -34,3 +34,41 @@ def test_huge_nanoseconds_overflow():
+        assert time1 - time2 == Timedelta("0 days 04:00:00")
+
+    def test_subtracts_dates_of_different_types(self):
+        """ Github issue 31793 """
```
do 
``` python
# GH31793
```
on the first line and put that on every function

arw2019 commented on pandas/tests/tslibs/test_timedeltas.py:
```diff
@@ -34,3 +34,41 @@ def test_huge_nanoseconds_overflow():
+        """ Github issue 31793 """
+
+        time1 = Timestamp("2020-10-28 22:08:00-04:00")
+        time2 = Timestamp("2020-10-28 22:08:00-00:00").to_pydatetime()
```
maybe just use the `datetime.datetime` constructor

AnjoMan replied:
ok, i am moving the tests to `tests/arithmetic/test_timedelta64.py` since most of this functionality is already covered here

arw2019 commented on pandas/tests/tslibs/test_timedeltas.py:
```diff
@@ -34,3 +34,41 @@ def test_huge_nanoseconds_overflow():
+        with pytest.raises(TypeError, match=msg):
+            time2 - time1
+
+    def test_subtracts_dates_of_different_timezones(self):
```
use `pytest.mark.parametrize` to write this and the test below as a single function. Parametrize on the two timestamps and the expected Timedelta

AnjoMan replied:
I don't really like this idea, i think each test covers a different piece of functionality:

1. can i subtract when the timezone offsets are not the same?
2. can i subtract when the type of each datetime is not the same?

to me, using parametrization would hide the intent of each test -- you have to read and comprehend 4 date-times and 2 time-deltas in order to understand what is different about the two cases being tested. this is explicit, you can read the test function name and right away you know why the test exists

arw2019 reviewed (changes requested):
Thanks @AnjoMan for the PR!

Some comments

arw2019 commented:
Also heads up about linting - needs to be fixed for the CI checks to pass

jbrockmendel commented on pandas/_libs/tslibs/timestamps.pyx:
```diff
@@ -368,9 +368,13 @@ cdef class _Timestamp(ABCTimestamp):
+            ]
+            if any(naive_timezones) and not all(naive_timezones):
+                raise TypeError("Cannot compute time difference between"
+                                " timezone-aware and naive datetime values.")
```
nitpick: put the space at the end of the previous line instead of the beginning of this line

AnjoMan replied:
yeah, the CI suggested this as well

jbrockmendel commented on pandas/tests/tslibs/test_timedeltas.py:
```diff
@@ -34,3 +34,41 @@ def test_huge_nanoseconds_overflow():
+
+
+class TestComputingDeltasWithTimezoneDifferences:
+    def test_fails_to_subtract_when_one_date_is_naive(self):
```
these should go in tests.scalar.timestamp.test_arithmetic

AnjoMan replied:
ok, it seems like there are also relevant tests in `tests/arithmatic/test_timedelta64.py`. i see a note in that file about needing to do deduplication and parametrization of the tests (oo, and it looks like you might have been the author of that note haha)

jbrockmendel commented:
When changing this for Timestamp, we'll need to make the same change for DatetimeArray

AnjoMan commented:
>  When changing this for Timestamp, we'll need to make the same change for DatetimeArray

@jbrockmendel thanks for the heads-up, I'll take a look and try to replicate this functionality there as well.

Just so we're clear, you are expecting to find the same bug when doing something like
```
date_array_1 = DatetimeArray(...)
date_array_2 = DatetimeArray(...)

date_array_1 - date_array_2
```
and one of the arrays has `datetime` objects instead of `Timestamp` objects?

jreback commented on pandas/tests/tslibs/test_timedeltas.py:
```diff
@@ -34,3 +34,41 @@ def test_huge_nanoseconds_overflow():
+    def test_fails_to_subtract_when_one_date_is_naive(self):
+        time1 = Timestamp("2020-10-28 22:08:00-04:00")
+        time2 = Timestamp("2020-10-28 22:08:00")
+
```
we already heavily test this

jreback reviewed (changes requested):
I -1 on adding different timezone subtraction operations. We do not allow this anywhere else. I think being explicit is a really good thing here.

AnjoMan commented on pandas/tests/tslibs/test_timedeltas.py:
```diff
@@ -34,3 +34,41 @@ def test_huge_nanoseconds_overflow():
+        assert time1 - time2 == Timedelta("0 days 04:00:00")
+        assert time2 - time1 == Timedelta("-1 days +20:00:00")
+
+    def test_subtracts_dates_that_are_both_naive(self):
```
a link to the relevant tests would be more helpful

pep8speaks commented:
Hello @AnjoMan! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`pandas/tests/arithmetic/test_timedelta64.py`](https://github.com/pandas-dev/pandas/blob/298431f8ef56dc1cd244cb751975208b48cd127f/pandas/tests/arithmetic/test_timedelta64.py):

> [Line 398:89](https://github.com/pandas-dev/pandas/blob/298431f8ef56dc1cd244cb751975208b48cd127f/pandas/tests/arithmetic/test_timedelta64.py#L398): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (92 > 88 characters)
> [Line 402:9](https://github.com/pandas-dev/pandas/blob/298431f8ef56dc1cd244cb751975208b48cd127f/pandas/tests/arithmetic/test_timedelta64.py#L402): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 403:89](https://github.com/pandas-dev/pandas/blob/298431f8ef56dc1cd244cb751975208b48cd127f/pandas/tests/arithmetic/test_timedelta64.py#L403): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (98 > 88 characters)

AnjoMan commented on pandas/tests/tslibs/test_timedeltas.py:
```diff
@@ -34,3 +34,41 @@ def test_huge_nanoseconds_overflow():
+
+    def test_subtracts_dates_of_different_timezones(self):
+        time1 = Timestamp("2020-10-28 22:08:00-04:00")
+        time2 = Timestamp("2020-10-28 22:08:00-00:00")
```
i'm having a hard time seeing what the can of worms will be here, can you elaborate?

to me, the complex part is determining the timezone if it is specified using a different library (e.g. `pytz`, `dateutil` and `datetime` all have different `implementations`). 

the actual subtraction part is just, you add the utc offset to the number of hours difference. pandas can already do it, and it would match the implementation of `datetime` which does support this operation.


Are there historical examples you can point to where this has been a significant problem?

jbrockmendel replied:
@jreback the simplest reason to allow this is because the stdlib datetime does

@Anjoman the the tzinfo library pytz/dateutil/timezone is irrelevant.  once you know that `not ((self.tzinfo is None) ^ (other.tzinfo is None))`, the difference is `Timedelta(self.value - other.value)` regardless.

jbrockmendel commented:
> Just so we're clear, you are expecting to find the same bug when doing something like
> ```
> date_array_1 = DatetimeArray(...)
> date_array_2 = DatetimeArray(...)
> date_array_1 - date_array_2
> ```
> and one of the arrays has datetime objects instead of Timestamp objects?

datetime objects vs Timestamp objects isn't really applicable here.  Think of DatetimeArray as being a vectorized Timestamp.  Just like with Timestamp, when subtracting `dta - dta2` we need to have `not ((dta.tzinfo is None) ^ (dta2.tzinfo is None))`.  ATM we also require that the tzinfos match, which this PR would cease to require.

jorisvandenbossche commented:
> [@jreback] I am -1 on adding different timezone subtraction operations. We do not allow this anywhere else. I think being explicit is a really good thing here.

Apart from the fact that the stdlib also supports this as @jbrockmendel mentioned, I also want to point out we actually do support differing timezones in several operations, eg comparison ops:

```
In [5]: t1 = pd.Timestamp("2012-01-01").tz_localize("UTC").tz_convert("Europe/Brussels")

In [6]: t2 = pd.Timestamp("2012-01-01").tz_localize("UTC").tz_convert("US/Eastern")

In [7]: t1 == t2
Out[7]: True

In [8]: t1 >= t2
Out[8]: True
```

arw2019 reviewed (commented):
In light of this discussions in #37605 does this PR go ahead? AFAICT yes but maybe I misunderstood @jbrockmendel @jorisvandenbossche @jreback @mroeschke 

If yes @AnjoMan you will need to merge master and add a whatsnew note. Not sure if this will get in in time for 1.2 - if not we'll have to wait until we tag the next major release (1.3 or 2.0)

jbrockmendel commented:
> In light of this discussions in #37605 does this PR go ahead? AFAICT yes but maybe I misunderstood

@jorisvandenbossche and I are +1 on this, @jreback was -1 though there's some evidence of miscommunication, and @mroeschke was -1 though I'm hoping he'll reconsider when looking at this context.

@arw2019 @AnjoMan I encourage you both to weigh in on #37605.

mroeschke commented:
I'm more -0 just in principle, but since we already support some mixed timezone aware operations, it's not a hill I will die on.

arw2019 commented:
> > In light of this discussions in #37605 does this PR go ahead? AFAICT yes but maybe I misunderstood
> 
> @jorisvandenbossche and I are +1 on this, @jreback was -1 though there's some evidence of miscommunication, and @mroeschke was -1 though I'm hoping he'll reconsider when looking at this context.
> 
> @arw2019 @AnjoMan I encourage you both to weigh in on #37605.

I'm +1 on adding this

jreback commented:
@AnjoMan if you can rebase we can take another look here

AnjoMan commented:
Ok, i've rebased on the current dev branch and added a "whatsnew" entry.

I noticed one of the CI checks fails with a [bad import](https://github.com/pandas-dev/pandas/pull/37329/checks?check_run_id=1876403833) -- it doesn't seem too related to what I've changed

![image](https://user-images.githubusercontent.com/2414360/107601671-645e6700-6bf5-11eb-8b5c-a11a2e9a925c.png)

arw2019 commented:
> Ok, i've rebased on the current dev branch and added a "whatsnew" entry.
> 
> I noticed one of the CI checks fails with a [bad import](https://github.com/pandas-dev/pandas/pull/37329/checks?check_run_id=1876403833) -- it doesn't seem too related to what I've changed
> 
> ![image](https://user-images.githubusercontent.com/2414360/107601671-645e6700-6bf5-11eb-8b5c-a11a2e9a925c.png)

will be fixed by #39744

systemcatch commented:
Hi is there anything further that needs to be done here?

AnjoMan commented:
> Hi is there anything further that needs to be done here?

Nothing else required on this AFAIK, other than @jreback has a review request that i think has been adressed

jbrockmendel commented on pandas/tests/arithmetic/test_timedelta64.py:
```diff
@@ -366,6 +367,8 @@ def test_subtraction_ops_with_tz(self):
         ts_tz2 = Timestamp("20130101").tz_localize("CET")
         dt_tz = ts_tz.to_pydatetime()
         td = Timedelta("1 days")
+        ts_utc = Timestamp("2020-10-22T22:00:00+00:00")
```
can you update this test to use box_with_array fixture to ensure we test all the relevant types

jbrockmendel commented on pandas/tests/scalar/timestamp/test_arithmetic.py:
```diff
@@ -99,7 +99,7 @@ def test_rsub_dtscalars(self, tz_naive_fixture):
             assert other.to_datetime64() - ts == td
         else:
-            msg = "subtraction must have"
+            msg = "Cannot subtract offset-naive and offset-aware datetimes."
```
should have tests for the changed Timestamp behavior (i.e. that is tested in this file, not in the file above)

AnjoMan replied:
ok, i can add tests here to cover subtraction behaviour of Timestamp. do you want coverage in both places? or is it enough just to test here?

jbrockmendel replied:
the scalar timestamp tests should be here.  the DatetimeArray/DatetimeIndex/Series[dt64] tests go in tests.arithmetic.test_datetime64

AnjoMan replied:
sounds good :+1:

systemcatch replied:
@jbrockmendel any further comments here?

AnjoMan replied:
Ok, i've moved the tests as requested

AnjoMan committed:
TST: add index type coverage to timedelta subtraction tests (#31793)

AnjoMan committed:
BUG: cannot subtract Timestamp with different timezones (#31793)

AnjoMan commented:
Ok, i think this is back on track -- there is one last failing check but it is a mypy issue that I'm fairly certain is not related to my changes at all:

![image](https://user-images.githubusercontent.com/2414360/110995112-cd95df00-8347-11eb-9fc1-2bec4fce266f.png)

jbrockmendel commented:
it seems likely that the mypy complaint is unrelated.  can you merge master and ill take another look

jbrockmendel commented on doc/source/whatsnew/v1.3.0.rst:
```diff
@@ -434,6 +434,7 @@ Timezones
 ^^^^^^^^^
 - Bug in different ``tzinfo`` objects representing UTC not being treated as equivalent (:issue:`39216`)
 - Bug in ``dateutil.tz.gettz("UTC")`` not being recognized as equivalent to other UTC-representing tzinfos (:issue:`39276`)
+- Bug in ``Timestamp`` and ``DatetimeIndex`` incorrectly raising a ``TypeError`` when subtracting two times that don't have the same timezone offset (:issue:`31793`)
```
"two times that don't have the same timezone offset" -> "two timezone-aware objects with mismatched timezones"

jbrockmendel commented on pandas/_libs/tslibs/timestamps.pyx:
```diff
@@ -319,10 +319,10 @@ cdef class _Timestamp(ABCTimestamp):
-                                "same timezones or no timezones")
+            if (self.tzinfo is None) ^ (other.tzinfo is None):
+                raise TypeError(
+                    "Cannot compare tz-naive and tz-aware datetime-like objects."
```
should this message be about subtraction?

AnjoMan replied:
Yup, makes sense! just so you know though, it won't be consistent across the board because in some of the vector implementations (e.g. `pandas/core/arrays/datetimes.py`) there is a timezone awareness check which is optimized, and when the check fails there is no way to know what operation was being done. so, some of the array / series based tests i've added will still assert 'cannot _compare_'

jbrockmendel commented on pandas/tests/arithmetic/test_datetime64.py:
```diff
@@ -864,6 +864,46 @@ def test_dt64arr_isub_timedeltalike_scalar(
+        expected = tm.box_expected(expected, box_with_array)
+        tm.assert_equal(result, expected)
+
+    def test_dt64_array_sub_dt64_array_with_different_timezone(self, box_with_array):
```
do we have tests somewhere in this file for the cases that still raise?

AnjoMan replied:
it appears not -- i'll add some tests here covering the error case

jbrockmendel commented on pandas/tests/series/methods/test_shift.py:
```diff
@@ -110,11 +110,10 @@ def test_shift(self, datetime_series):
-        with pytest.raises(TypeError, match=msg):
-            s - s2
+        exp = Series(TimedeltaIndex(["6 hours"] * 5), name="foo")
+        result = s - s2
```
assert something about this?

simonjayhawkins commented:
@AnjoMan can you address @jbrockmendel comments

AnjoMan commented:
I've updated the PR to address feedback from @jbrockmendel 

1. In `pandas/_libs/tslibs/timestamps.pyx` we raise 'cannot _subtract_' instead of 'cannot _compare_' as it is more specific (in other cases we still raise 'cannot _compare_' as array / series implementations of __sub__ have an awareness check that is optimized for vectors of timestamps
2. add a missing assertion in `pandas/tests/series/methods/test_shift.py `
3. add coverage in `pandas/tests/arithmetic/test_datetime64.py` of the failure case when subtracting between series/array and series/array/timestamp
4. reworded whatsnew.rst entry

jbrockmendel commented on pandas/tests/arithmetic/test_datetime64.py:
```diff
@@ -864,6 +864,61 @@ def test_dt64arr_isub_timedeltalike_scalar(
+        expected = tm.box_expected(expected, box_with_array)
+        tm.assert_equal(result, expected)
+
+        msg = "Cannot compare tz-naive and tz-aware datetime-like objects"
```
does this message need to be updated?

AnjoMan replied:
i added a try/catch to reraise this as subtract in the case where this was coming up as 'compare'

jbrockmendel commented on pandas/tests/arithmetic/test_datetime64.py:
```diff
@@ -864,6 +864,61 @@ def test_dt64arr_isub_timedeltalike_scalar(
+        expected = tm.box_expected(expected, box_with_array)
+        tm.assert_equal(result, expected)
+
+        msg = "Cannot compare tz-naive and tz-aware datetime-like objects"
```
needs updating?

jbrockmendel commented on pandas/tests/arithmetic/test_timedelta64.py:
```diff
@@ -384,60 +386,60 @@ def _check(result, expected):
         expected = Timedelta("0 days")
         _check(result, expected)
 
+        # gotta chedk this
```
not sure what this comment is adding

jbrockmendel commented on pandas/tests/scalar/timestamp/test_arithmetic.py:
```diff
@@ -109,6 +110,37 @@ def test_timestamp_sub_datetime(self):
+        assert isinstance(result, Timedelta)
+        assert result == Timedelta("0 days 06:00:00")
+
+    def test_subtracting_involving_datetime_with_different_tz(self):
```
could use tz_aware_fixture

AnjoMan replied:
it seems like we need to modify the fixture to pair an expectation for the offset, if we want to do this. Wondering what concern you want to address with this? It seems like there are two different things to potentiallly cover:

1. varying the format in which the offset is specified (e.g. `US/Eastern`, `Fixedoffset(-250)`, `timezone(timedelta(hours-5))`) to assert all result in the same offset
2. varying the timezone to make sure we can compute the offset implied (e.g. "US/Eastern", "US/Pacific", "Asia/Tokyo")

AnjoMan replied:
I added a fixture for capturing the expected time difference, and wrote a test to assert we can compute the right value

AnjoMan committed:
TST: 31739 - add coverage of subtracting datetimes w/differing timezones

jbrockmendel commented on pandas/conftest.py:
```diff
@@ -1088,6 +1092,15 @@ def tz_aware_fixture(request):
     return request.param
 
 
+@pytest.fixture()
```
i think the parens here unnecessary?

jbrockmendel commented on pandas/core/arrays/datetimes.py:
```diff
@@ -764,11 +762,10 @@ def _sub_datetimelike_scalar(self, other):
-            )
+        try:
+            self._assert_tzawareness_compat(other)
+        except TypeError as e:
```
can you make this "err" instead of "e"; 1-letter variable names make grepping tough

AnjoMan replied:
:+1:

jbrockmendel commented on pandas/core/arrays/datetimes.py:
```diff
@@ -713,12 +713,10 @@ def _sub_datetime_arraylike(self, other):
+        try:
+            self._assert_tzawareness_compat(other)
+        except TypeError as e:
+            raise TypeError(str(e).replace("compare", "subtract"))
```
instead of str.replace, can you edit the existing message that keeps `type(self).__name__`

AnjoMan replied:
i did:
`raise type(error)(new_message) from error`

so we can support subclasses and get the stack trace. the alternative to this is to modify the 'args' of the existing error, but that seems like it would depend on the error being raised in the right way

jreback commented:
in theory ok with this, if you can merge master and will look

AnjoMan commented:
@jreback i updated it

jbrockmendel reviewed (approved):
LGTM cc @jreback

jbrockmendel commented:
@AnjoMan can you merge master one more time

AnjoMan commented:
@jbrockmendel i've updated the branch against pandas master

jbrockmendel commented:
gentle ping @jreback

jreback commented:
ok on board here. @AnjoMan one more rebase if you would and @mroeschke if any comments.

AnjoMan commented:
i've rebased on top of c3d335735d9f201910ddef709af3fbc55252bec2

mroeschke commented:
Thanks for hanging in there @AnjoMan. Great work!

jbrockmendel commented:
nice!

AnjoMan commented:
thanks @jbrockmendel @jreback @mroeschke for all the feedback and help getting this together!
